### PR TITLE
EmailServer Bee

### PR DESCRIPTION
**Pull request by mkrauser**
Hi,

this bee creates an SMTP-Server and triggers an event when an eMail is received. [flashmob/go-guerilla](https://github.com/flashmob/go-guerrilla) is used as SMTP-Implementation. TLS is fully supported. The event contains some meta-data and the (raw-)body of the mail.

There is one more change, I need to explain:
The logging outputs all properties of all events. If an eMail is a bit longer or contains attachments, this output will be very big. Therefore I've limited the log-output of event-options to 1000 characters, like shown below. 
This applies to all bees, but it should only affekt very few events. I hope this change is OK. 

```
INFO[0072]                                              
INFO[0072] Event received: smtp /  -                    
INFO[0072] 	Options: {sender string info@krauser.eu}    
INFO[0072] 	Options: {remote_ip string 127.0.0.1}       
INFO[0072] 	Options: {recipients string[] [mk@some-test-host.com]} 
INFO[0072] 	Options: {subject string This is the subject} 
INFO[0072] 	Options: {tls boolean true}                 
INFO[0072] 	Options: {headers string[] map[Content-Type:[multipart/alternative; boundary="Apple-Mail=_2A2729D5-AC21-4979-A230-EE34509542AA"] Date:[Fri, 3 May 2019 21:47:12 +0200] From:[Matthias Krauser <info@some-test-host.com>] Message-Id:[<BA204D5C-942B-4E49-8756-2B7A94A6B05E@some-test-host.com>] Mime-Version:[1.0 (Mac OS X Mail 12.4 \(3445.104.8\))] Subject:[This is the subject] To:[mk@some-test-host.com] X-Mailer:[Apple Mail (2.3445.104.8)]]} 
INFO[0072] 	Options: {body string From: Matthias Krauser <info@some-test-host.com>
Content-Type: multipart/alternative;
	boundary="Apple-Mail=_2A2729D5-AC21-4979-A230-EE34509542AA"
Mime-Version: 1.0 (Mac OS X Mail 12.4 \(3445.104.8\))
Subject: This is the subject
Message-Id: <BA204D5C-942B-4E49-8756-2B7A94A6B05E@krauser.eu>
Date: Fri, 3 May 2019 21:47:12 +0200
To: mk@dev.krauser.eu
X-Mailer: Apple Mail (2.3445.104.8)


--Apple-Mail=_2A2729D5-AC21-4979-A230-EE34509542AA
Content-Transfer-Encoding: 7bit
Content-Type: text/plain;
	charset=us-ascii

Test with an attachment

--Apple-Mail=_2A2729D5-AC21-4979-A230-EE34509542AA
Content-Type: multipart/related;
	type="text/html";
	boundary="Apple-Mail=_7F98C403-54B9-451D-A084-8F7E920FFAFD"


--Apple-Mail=_7F98C403-54B9-451D-A084-8F7E920FFAFD
Content-Transfer-Encoding: 7bit
Content-Type: text/html;
	charset=us-ascii

<html><head><meta http-equiv="Content-Type" content="text/html; charset=us-ascii"></head><body style="word-wrap: break-word; -webkit-nbsp-mode: space; line-b... (2870524 more characters) 
```

### CODE REVIEW

File: bees/emailserverbee/emailserverbee.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		return
+	}
+}
+
+// ReloadOptions parses the config options and initializes the Bee.
+func (mod *EmailServerBee) ReloadOptions(options bees.BeeOptions) {
+	mod.SetOptions(options)
+
+	mod.Logln("ReloadOptions")
+	mod.Logf("#%v\n", mod)
```
Comment (CalmBit) [inline bees/emailserverbee/emailserverbee.go]: Debug logging here? Might be good to remove before merge 😄

File: bees/emailserverbee/emailserverbee.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	options.Bind("allowedHosts", &mod.allowedHosts)
+	options.Bind("startTLSOn", &mod.startTLSOn)
+	options.Bind("tlsAlwaysOn", &mod.tlsAlwaysOn)
+	options.Bind("privateKeyFile", &mod.privateKeyFile)
+	options.Bind("publicKeyFile", &mod.publicKeyFile)
+	options.Bind("hostname", &mod.hostname)
+	options.Bind("maxSize", &mod.maxSize)
+	options.Bind("timeout", &mod.timeout)
+	options.Bind("maxClient", &mod.maxClients)
+	mod.Logf("#%v\n", mod)
```
Comment (CalmBit) [inline bees/emailserverbee/emailserverbee.go]: Ditto!

File: bees/emailserverbee/emailserverbee.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+									Value: e.Subject,
+								},
+								{
+									Name:  "tls",
+									Type:  "boolean",
+									Value: e.TLS,
+								},
+								{
+									Name:  "headers",
+									Type:  "string[]",
```
Comment (muesli) [inline bees/emailserverbee/emailserverbee.go]: This should read `[]string`, as we decided to stick with Go's conventions. It currently hasn't much of an effect within Beehive, but the admin interface will make use of that information, soon.

File: bees/emailserverbee/emailserverbee.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+									Value: e.MailFrom.String(),
+								},
+								{
+									Name:  "remote_ip",
+									Type:  "string",
+									Value: e.RemoteIP,
+								},
+								{
+									Name:  "recipients",
+									Type:  "string[]",
```
Comment (muesli) [inline bees/emailserverbee/emailserverbee.go]: This should read []string, as we decided to stick with Go's conventions. It currently hasn't much of an effect within Beehive, but the admin interface will make use of that information, soon.

File: bees/emailserverbee/emailserverbeefactory.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+				},
+				{
+					Name:        "remote_ip",
+					Description: "The IP from the remote-server",
+					Type:        "string",
+				},
+				{
+					Name:        "recipients",
+					Description: "The recipients",
+					Type:        "string[]",
```
Comment (muesli) [inline bees/emailserverbee/emailserverbeefactory.go]: This should read []string, as we decided to stick with Go's conventions. It currently hasn't much of an effect within Beehive, but the admin interface will make use of that information, soon.

File: bees/emailserverbee/emailserverbeefactory.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+				},
+				{
+					Name:        "tls",
+					Description: "Indicates, if the mail was transmitted using tls",
+					Type:        "bool",
+				},
+				{
+					Name:        "headers",
+					Description: "The Email headers",
+					Type:        "string",
```
Comment (muesli) [inline bees/emailserverbee/emailserverbeefactory.go]: Should be a `[]string` as it is used when emitting events.

**Comment by muesli** [comment]
Nice work & a really interesting bee you wrote there! In addition to @CalmBit's comments, I just added a few change requests, all revolving around the `[]string` type.

Looks good to me, otherwise!

File: bees/events.go [hunk 7]
```diff
@@ -51,7 +52,8 @@ func handleEvents() {
 		log.Println()
 		log.Println("Event received:", event.Bee, "/", event.Name, "-", GetEventDescriptor(&event).Description)
 		for _, v := range event.Options {
-			log.Println("\tOptions:", v)
+			vv := truncateString(fmt.Sprintln(v), 1000)
```
Comment (muesli) [inline bees/events.go]: I like the change, but as a general advice I would recommend moving unrelated changes into a separate commit & PR, to keep them as granular & atomic as possible. This makes it a lot easier to maintain code in the long run.

**Comment by mkrauser** [comment]
@muesli @CalmBit  Thanks four your review guys! The issues have been fixed

File: bees/events.go (hunk 7 above)
Comment (mkrauser) [reply bees/events.go]: Thanks, I'll keep that in mind

**Comment by muesli** [comment]
@mkrauser Just a heads up: you seem to have overlooked two `string[]` in `emailserverbee.go` still.

**Comment by mkrauser** [comment]
Sorry, one should never correct a PR in a hurry ;-)

**Comment by muesli** [comment]
Thank you for your contribution to Beehive, @mkrauser!
